### PR TITLE
Add a migration document

### PR DESCRIPTION
**Pull request by voltur01**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -137,6 +137,10 @@ $ clang \
 -o example example.c
 ```
 
+See [Migrating from Arm GNU Toolchain](https://github.com/ARM-software/LLVM-embedded-toolchain-for-Arm/blob/main/docs/migrating.md)
+for advise on using the LLVM Embedded Toolchain with existing projects
```
Comment (mplatings) [inline README.md:141]:
```suggestion
for advice on using the LLVM Embedded Toolchain with existing projects
```

File: docs/migrating.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
+# Migrating from Arm GNU Toolchain
+
+## Overview
+
+[Arm GNU Toolchain](https://developer.arm.com/Tools%20and%20Software/GNU%20Toolchain)
+is the GNU Toolchain for the Arm Architecture released by Arm and traditionally
+used for embedded development.
+
+Generally LLVM toolchain tries to be a drop in replacement for the GNU toolchain,
```
Comment (mplatings) [inline docs/migrating.md:9]:
```suggestion
Generally the LLVM toolchain tries to be a drop in replacement for the GNU toolchain,
```

File: docs/migrating.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+Some known differences and migration strategies are summarized below.
+
+## Benefits
+
+Using multiple toolchains to build a project benefits from different checks
+and warnings present in different compilers to catch more issues,
+particularly C and C++ standards compliance, during build time.
+
+LLVM Embedded Toolchain provides support for multiple sanitizers
```
Comment (mplatings) [inline docs/migrating.md:20]:
```suggestion
LLVM Embedded Toolchain for Arm provides support for multiple sanitizers
```
I think it's less ambiguous if we use the proper name.

File: docs/migrating.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+and memory safety features to also catch typical issues at runtime during testing,
+see the [LLVM documentation](https://clang.llvm.org/docs/index.html)
+and [samples](https://github.com/ARM-software/LLVM-embedded-toolchain-for-Arm/tree/main/samples/src).
+
+LLVM Embedded Toolchain provides superior performance when targeting the
```
Comment (mplatings) [inline docs/migrating.md:25]:
```suggestion
LLVM Embedded Toolchain for Arm provides superior performance when targeting the
```

File: docs/migrating.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+## Benefits
+
+Using multiple toolchains to build a project benefits from different checks
+and warnings present in different compilers to catch more issues,
+particularly C and C++ standards compliance, during build time.
+
+LLVM Embedded Toolchain provides support for multiple sanitizers
+and memory safety features to also catch typical issues at runtime during testing,
+see the [LLVM documentation](https://clang.llvm.org/docs/index.html)
```
Comment (mplatings) [inline docs/migrating.md:22]:
```suggestion
see the [Clang documentation](https://clang.llvm.org/docs/index.html)
```

File: docs/migrating.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+|Version number|GNU macro|LLVM macro|
+|-------|---------|----------|
+|Major|`__GNUC__`|`__clang_major__`|
+|Minor|`__GNUC_MINOR__`|`__clang_minor__`|
+|Patch level|`__GNUC_PATCHLEVEL__`|`__clang_patchlevel__`|
+
+## C and C++ language extensions
+
+LLVM supports majority of GNU C and C++ extensions as described in
```
Comment (mplatings) [inline docs/migrating.md:56]:
```suggestion
LLVM supports the majority of GNU C and C++ extensions as described in
```

File: docs/migrating.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+accepted by both.
+
+For example, GNU and LLVM compilers differ in:
+* Use of `.N` and `.W` suffixes for Thumb instructions.
+* Use of `S` suffix for flag setting instructions.
+
+## Multilib support
+
+LLVM toolchain provides multilib support similar to the GNU toolchain,
+see _Using the toolchain_ section in the [README](https://github.com/ARM-software/LLVM-embedded-toolchain-for-Arm/blob/main/README.md),
```
Comment (mplatings) [inline docs/migrating.md:82]:
```suggestion
see _Using the toolchain_ section in the [README](https://github.com/ARM-software/LLVM-embedded-toolchain-for-Arm/blob/main/README.md#using-the-toolchain),
```

File: docs/migrating.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+|Use case|GNU options|LLVM options|
+|--------|-----------|------------|
+|No semihosting|`--specs=nosys.specs`|`-lcrt0`|
+|Semihosting|`--specs=rdimon.specs`|`-lcrt0-semihost -lsemihost`|
+|Newlib-nano|`--specs=nano.specs`|Not available: `picolibc` is a derivative of `newlib-nano`.
+
+## Linker
+
+`lld` is designed as a drop in replacement for the GNU `arm-none-eabi-ld`,
```
Comment (mplatings) [inline docs/migrating.md:93]:
```suggestion
`lld` is designed as a drop in replacement for GNU `ld`,
```

File: docs/migrating.md [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+however there are some known differences to take into account, see:
+* [Linker Script implementation notes and policy](https://lld.llvm.org/ELF/linker_script.html)
+in the LLVM documentation.
```
Comment (mplatings) [inline docs/migrating.md:96]:
```suggestion
in the LLD documentation.
```

File: docs/migrating.md [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+* [LLD and GNU linker incompatibilities](https://maskray.me/blog/2020-12-19-lld-and-gnu-linker-incompatibilities)
```
Comment (mplatings) [inline docs/migrating.md:97]:
```suggestion
* The [LLD and GNU linker incompatibilities](https://maskray.me/blog/2020-12-19-lld-and-gnu-linker-incompatibilities)
```

File: docs/migrating.md [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+the application interrupt vector table should be placed into the `init`
+[linker script](https://github.com/picolibc/picolibc/blob/main/doc/linking.md)
+section and referenced by the `__interrupt_vector` symbol.
+
+## Standard input and output
+
+See [Picolibc and Operating Systems](https://github.com/picolibc/picolibc/blob/main/doc/os.md)
+for the details on redirecting `stdin`, `stdout` and `stderr`.
+
+`baremetal-uart` [sample](https://github.com/ARM-software/LLVM-embedded-toolchain-for-Arm/tree/main/samples/src/baremetal-uart)
```
Comment (mplatings) [inline docs/migrating.md:115]:
```suggestion
The `baremetal-uart` [sample](https://github.com/ARM-software/LLVM-embedded-toolchain-for-Arm/tree/main/samples/src/baremetal-uart)
```

**Review by mplatings — COMMENTED** [review]
Looks very helpful overall, just some grammatical nits.

File: docs/migrating.md [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+LLVM toolchain provides multilib support similar to the GNU toolchain,
+see _Using the toolchain_ section in the [README](https://github.com/ARM-software/LLVM-embedded-toolchain-for-Arm/blob/main/README.md),
+however uses different command line options to control selection of semihosting.
+
+|Use case|GNU options|LLVM options|
+|--------|-----------|------------|
+|No semihosting|`--specs=nosys.specs`|`-lcrt0`|
+|Semihosting|`--specs=rdimon.specs`|`-lcrt0-semihost -lsemihost`|
+|Newlib-nano|`--specs=nano.specs`|Not available: `picolibc` is a derivative of `newlib-nano`.
```
Comment (smithp35) [inline docs/migrating.md:89]: I'm not sure whether it is a derivative, I think it is closer to a mashup between AVR C (tinystdio) and newlib . On further study I think it is closer to being an "equivalent" of newlib nano.

File: docs/migrating.md [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+## Toolchain identification
+
+Toolchain version macros:
+
+|Version number|GNU macro|LLVM macro|
+|-------|---------|----------|
+|Major|`__GNUC__`|`__clang_major__`|
+|Minor|`__GNUC_MINOR__`|`__clang_minor__`|
+|Patch level|`__GNUC_PATCHLEVEL__`|`__clang_patchlevel__`|
+
```
Comment (smithp35) [inline docs/migrating.md:53]: Do we need some more explanation of the table?

Clang will define all of these, using 4, 2 and 1 (a very old GCC!) for the GNU defines. This is OK for code that checks for the existence, but not good for code that tries to detect a recent GCC to infer whether certain builtins exist.

**Review by smithp35 — COMMENTED** [review]
A couple of small comments, but overall looks good.

File: docs/migrating.md (hunk 12 above)
Comment (voltur01) [reply docs/migrating.md]: Agreed, replaced.

File: docs/migrating.md (hunk 13 above)
Comment (voltur01) [reply docs/migrating.md:53]: Very good point - this is an important detail, added.

File: docs/migrating.md [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+Note that `clang` defines GNU macros for compatibility too:
+`__GNUC__` equal to `4`, `__GNUC_MINOR__` equal to `2`,
+and `__GNUC_PATCHLEVEL__` equal to `1`.
+
+## C and C++ language extensions
+
+LLVM supports the majority of GNU C and C++ extensions as described in
```
Comment (mplatings) [inline docs/migrating.md:60]:
```suggestion
Clang supports the majority of GNU C and C++ extensions as described in
```

File: README.md [hunk 15]
```diff
@@ -137,6 +137,10 @@ $ clang \
 -o example example.c
 ```
 
+See [Migrating from Arm GNU Toolchain](https://github.com/ARM-software/LLVM-embedded-toolchain-for-Arm/blob/main/docs/migrating.md)
+for advice on using the LLVM Embedded Toolchain with existing projects
```
Comment (mplatings) [inline README.md:141]:
```suggestion
for advice on using LLVM Embedded Toolchain for Arm with existing projects
```

File: docs/migrating.md [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+compilers, however most of the time it is possible to write code that is
+accepted by both.
+
+For example, GNU and LLVM compilers differ in:
+* Use of `.N` and `.W` suffixes for Thumb instructions.
+* Use of `S` suffix for flag setting instructions.
+
+## Multilib support
+
+LLVM toolchain provides multilib support similar to the GNU toolchain,
```
Comment (mplatings) [inline docs/migrating.md:85]:
```suggestion
Clang provides multilib support similar to the GNU toolchain,
```
Or maybe
```suggestion
LLVM Embedded Toolchain for Arm provides multilib support similar to the GNU toolchain,
```

**Review by mplatings — COMMENTED** [review]
Sorry a few more nits

**Review by smithp35 — APPROVED** [review]
Thanks for the update. I've left my approval. Happy once all the other reviewers are.

**Review by mplatings — APPROVED** [review]

**Comment by voltur01** [comment]
Thank you both for the review and good suggestions!

**Pull request merged**